`datatypes.Queue` can't be assigned to `shared` variable

### Describe the bug

When trying to create a new empty `datatypes.Queue` instance and assign it to a variable that is marked as `shared` - the compiler produces an error, that is potentially related to cgen error

### Reproduction Steps

```
import datatypes

shared test := datatypes.Queue[int]{}
```

### Expected Behavior

Compiled without error

### Current Behavior

```
/tmp/v_1000/seven-concurrency-models-v.5538685564126413957.tmp.c:28753: error: '{' expected (got ";")
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.2 f54f156.f82529e

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.2 f54f156.f82529e
OS: linux, "Garuda Linux"
Processor: 16 cpus, 64bit, little endian, AMD Ryzen 7 3800X 8-Core Processor

getwd: /home/yuart/Projects/seven-concurrency-models-v
vexe: /home/yuart/Projects/v/v
vexe mtime: 2023-11-05 12:24:30

vroot: OK, value: /home/yuart/Projects/v
VMODULES: OK, value: /home/yuart/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.42.1
Git vroot status: weekly.2023.42-147-gf82529e0 (15 commit(s) behind V master)
.git/config present: true

CC version: cc (GCC) 13.2.1 20230801
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -570,11 +570,18 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 							if val_sym.info.generic_types.len > 0 {
 								if val is ast.StructInit {
 									var_styp := g.typ(val.typ)
-									g.write('${var_styp} ')
+									if var_type.has_flag(.shared_f) {
+										g.write('__shared__${var_styp}* ')
+									} else {
+										g.write('${var_styp} ')
+									}
 									is_used_var_styp = true
 								} else if val is ast.PrefixExpr {
 									if val.op == .amp && val.right is ast.StructInit {
 										var_styp := g.typ(val.right.typ.ref())
+										if var_type.has_flag(.shared_f) {
+											g.write('__shared__')
+										}
 										g.write('${var_styp} ')
 										is_used_var_styp = true
 									}

--- a/vlib/v/tests/shared_generic_test.v
+++ b/vlib/v/tests/shared_generic_test.v
@@ -1,0 +1,14 @@
+struct Foo[T] {
+	a T
+}
+
+fn test_shared_struct_has_generics() {
+	shared foo := Foo[int]{1}
+	lock foo {
+		assert foo.a == 1
+	}
+	shared bar := &Foo[int]{1}
+	lock bar {
+		assert bar.a == 1
+	}
+}

